doc/dev/osd_internals/partial_object_recovery.rst

```diff
@@ -0,0 +1,121 @@
+max_num_intervals is an upbound of the number of intervals in clean_offsets
+so that the memory cost of clean_offsets is always bounded.
+The shortest clean interval will be trimmed if the number of intervals
+in clean_offsets exceeds the boundary.
+    etc. max_num_intervals=2, clean_offsets:{[5~10], [20~5]}
+    then new interval [30~10] will evict out the shortest one [20~5]
+    finally, clean_offsets becomes {[5~10], [30~10]}
+
+
+Procedures for Partial Object Recovery
+
+Firstly, OpContext and pg_log_entry_t should contain ObjectCleanRegion.
+In do_osd_ops(), finish_copyfrom(), finish_promote(), corresponding content
+in ObjectCleanRegion should mark dirty so that trace the modification of an object.
+Also update ObjectCleanRegion in OpContext to its pg_log_entry_t.
+
+Secondly, pg_missing_set can build and rebuild correctly.
+when calculating pg_missing_set during peering process,
+also merge ObjectCleanRegion in each pg_log_entry_t.
+    etc. object aa has pg_log:
+        26'101 {[0~4096, 8192~MAX], false}
+        26'104 {0~8192, 12288~MAX, false}
+        28'108 {[0~12288, 16384~MAX], true}
+    missing_set for object aa: merge pg_log above --> {[0~4096, 16384~MAX], true}
+    which means 4096~16384 is modified and object_map is also modified on version 28'108
+Also, OSD may be crash after merge log.
+Therefore, we need to read_log and rebuild pg_missing_set. For example, pg_log is
+    object aa: 26'101 {[0~4096, 8192~MAX], false}
+    object bb: 26'102 {[0~4096, 8192~MAX], false}
+    object cc: 26'103 {[0~4096, 8192~MAX], false}
+    object aa: 26'104 {0~8192, 12288~MAX, false}
+    object dd: 26'105 {[0~4096, 8192~MAX], false}
+    object aa: 28'108 {[0~12288, 16384~MAX], true}
+Orignally, if bb,cc,dd is recovered, and aa is not.
+So we need to rebuild pg_missing_set for object aa,
+and find aa is modified on version 28'108.
+If version in object_info is 26'96 < 28'108,
+we don't need to consider 26'104 and 26'101 because the whole object will be recovered.
+However, Partial Object Recovery should also require us to rebuild ObjectCleanRegion.
+Knowing whether the object is modified is not enough.
+Therefore, we also need to tranverse the pg_log before,
+that says 26'104 and 26'101 also > object_info(26'96)
+and rebuild pg_missing_set for object aa based on those three logs: 28'108, 26'104, 26'101.
+The way how to merge logs is the same as mentioned above
+
+Finally, finish the push and pull process based on pg_missing_set.
+Updating copy_subset in recovery_info based on ObjectCleanRegion in pg_missing_set.
+copy_subset indicates the intervals of content need to pull and push.
+The complicated part here is submit_push_data
+and serval cases should be considered seperately.
+what we need to consider is how to deal with the object data,
+object data makes up of omap_header, xattrs, omap, data:
+
+case 1 -- first && complete: since object recovering is finished in a single PushOp,
```

Comment:
can you update this based on the current implementation? in particular there's no cloning of omap anymore, I'm not sure if everything else is still the same